text/logs/0114-log-correlation.md

```diff
@@ -0,0 +1,149 @@
+# Conventions for Trace and Resource Association in Logs
+
+Create standards for correlating traces and resources in existing text logs.
+
+## Motivation
+
+Traces and logs present two separate perspectives on what was occurring when an
+operation was executed. In order to tie together the two perspectives, certain
+information needs to be added to the logs when they were generated within an
+open span. This document lays out the information that needs to be emitted in
+order to correlate the two data types, and how the data can be added to existing
+text log formats so that it can be recognized when the logs are parsed.
+
+## Explanation
+
+Two types of correlation need to happen to tie a log record into its full
+execution context. The first is Request Correlation, which ties the record to
+operations that were occurring when the record was created. The second is
+Resource Correlation, which associates the entry with where the event occurred
+such as a host, a pod, or a virtual machine.
+
+### Request Correlation
+
+Request correlation is achieved primarily with two values, a trace identifier
+and a span identifier. A trace may contain multiple spans, arranged as a tree,
+and may also contain links to other related spans. The combination of a trace
+identifier and a span identifier corresponds to a specific scope of work. In the
+tracing API, that scope can also contain attributes and events that describe
+what work the program being traced was currently performing.
+
+In most cases when traces and logs are used in tandem, the attributes of the
+current span do not need to be added to the log entry, because it would duplicate
+transmission of those values. As one span is likely associated with several (or
+many) log entries, it is more efficient to transmit span attributes with the
+span once rather than many times with each log entry. As a result, for most
+purposes the goal is to set the three values of traceid, spanid, and traceflags
+into each log entry that should be associated with that span.
+
+Request correlation fields only make sense when the log event occurred within
+an open span, so the fields in the table below are only required when the log
+event is to be correlated.
+
+| Field      | Required | Format
+| :--------- | :------- | :--------------------------------------------------
+| traceid    | Yes      | 16-byte numeric value as Base16-encoded string
```

Comment:
At least in Python, the log format requires all fields to be present on the log record object or it throws an exception. Also, the log format cannot be changes dynamically depending on whether an active span (or fields on log record) is present or not. So in this case we can either leave the fields empty so that the formatted logs look like `trace_id= span_id= ` or we can set them to zero so it looks like `trace_id=0 span_id=0`. Can we clarify what values the fields should be set to in case no active span is found in the context?

Reply:
@owais what would you prefer? Also, it seems like a fairly specific use case, could it be left to whoever configures the logger on how they want to represent the lack of trace context?

Reply:
I have a slight preference for `0`.

>  could it be left to whoever configures the logger on how they want to represent the lack of trace context?

It could but then backends will have to deal with multiple ways of representing nil values (0 vs '' vs null in JSON, etc). It would be nicer if the spec recommended one thing when the field cannot be omitted completely so analysis tools can have a standard way of detecting it.